src/PaymentServer/Persistence.hs

```diff
@@ -64,42 +72,87 @@ class VoucherDatabase d where
     -> Fingerprint                -- ^ The retry-enabling fingerprint for this redemption
     -> IO (Either RedeemError ()) -- ^ Left indicating the redemption is not allowed or Right indicating it is.
 
--- | MemoryVoucherDatabase is a voucher database that only persists state
--- in-memory.  The state does not outlive the process which creates it (nor
--- even the MemoryVoucherDatabase value).  This is primarily useful for
--- testing.
-data MemoryVoucherDatabase =
-  Memory {
+-- | VoucherDatabaseState is a type that captures whether we are using an
+-- in-memory voucher database that only persists state in-memory or
+-- a persistent one that writes to SQLite database.  The in-memory database
+-- state does not outlive the process which creates it (nor even the
+-- VoucherDatabase value). This is primarily useful for testing.
+-- `SQLiteDB` is useful for the production use where the state needs to persist.
+data VoucherDatabaseState =
+  MemoryDB {
     -- | A set of vouchers which have been paid for.
     paid :: IORef (Set.Set Voucher)
     -- | A mapping from redeemed vouchers to fingerprints associated with the
     -- redemption.
   , redeemed :: IORef (Map.Map Voucher Fingerprint)
   }
+  | SQLiteDB { conn :: Sqlite.Connection }
 
-instance VoucherDatabase MemoryVoucherDatabase where
-  payForVoucher Memory{ paid = paid, redeemed = redeemed } voucher = do
+instance VoucherDatabase VoucherDatabaseState where
+  payForVoucher MemoryDB{ paid = paid, redeemed = redeemed } voucher =
     modifyIORef paid (Set.insert voucher)
-    return ()
 
-  redeemVoucher Memory{ paid = paid, redeemed = redeemed } voucher fingerprint = do
+  payForVoucher SQLiteDB{ conn = conn } voucher =
+    insertVoucher conn voucher
+
+  redeemVoucher MemoryDB{ paid = paid, redeemed = redeemed } voucher fingerprint = do
     unpaid <- Set.notMember voucher <$> readIORef paid
     existingFingerprint <- Map.lookup voucher <$> readIORef redeemed
-    case (unpaid, existingFingerprint) of
-      (True, _) ->
-        return $ Left NotPaid
-      (False, Nothing) -> do
-        modifyIORef redeemed (Map.insert voucher fingerprint)
+    let insertFn voucher fingerprint = modifyIORef redeemed (Map.insert voucher fingerprint)
+    redeemVoucherHelper (unpaid, existingFingerprint) voucher fingerprint insertFn
+
+  redeemVoucher SQLiteDB { conn = conn } voucher fingerprint = do
+    unpaid <- isVoucherUnpaid conn voucher
+    existingFingerprint <- listToMaybe <$> getVoucherFingerprint conn voucher
+    let insertFn = insertVoucherAndFingerprint conn
+    redeemVoucherHelper (unpaid, existingFingerprint) voucher fingerprint insertFn
```

Comment:
Something in this function (I think) causes this:

```
ConversionFailed {errSQLType = "INTEGER", errHaskellType = "Text", errMessage = "need a text"}
```

but I wasn't easily able to track down what.  Can we improve the error handling/reporting around the SQLite3 operations?

Also, I think this function might need to be executed in the context of a SQLite3 transaction?  Probably a read isolated transaction.  There is only one write and it happens at the end but if a client tries to redeem one voucher twice "at the same time" then this code might see it as unused both times and then try to issue tokens for it twice.

`withExclusiveTransaction` perhaps?

Reply:
I just pushed a change to use `withExclusiveTransaction`.